master/internal/trial.go

```diff
@@ -643,62 +614,28 @@ func (t *trial) processAllocated(
 
 	ctx.Log().Infof("starting trial container: %v", w)
 
-	additionalFiles := archive.Archive{
-		t.agentUserGroup.OwnedArchiveItem(
-			trialEntrypointFile,
-			etc.MustStaticFile(etc.TrialEntrypointScriptResource),
-			trialEntrypointMode,
-			tar.TypeReg,
-		),
-
-		t.agentUserGroup.OwnedArchiveItem(trialSSHDir, nil, trialSSHDirMode, tar.TypeDir),
-		t.agentUserGroup.OwnedArchiveItem(trialAuthorizedKeysFile,
-			t.publicKey,
-			trialAuthorizedKeysMode,
-			tar.TypeReg,
-		),
-		t.agentUserGroup.OwnedArchiveItem(
-			trialRSAPublicKeyFile, t.publicKey, trialRSAPublicKeyMode, tar.TypeReg,
-		),
-		t.agentUserGroup.OwnedArchiveItem(
-			trialRSAPrivateKeyFile, t.privateKey, trialRSAPrivateKeyMode, tar.TypeReg,
-		),
-		t.agentUserGroup.OwnedArchiveItem(trialSSHDConfigFile,
-			etc.MustStaticFile(etc.SSHDConfigResource),
-			trialSSHDConfigMode,
-			tar.TypeReg,
-		),
-
-		archive.RootItem(
-			trialSSHConfigFile,
-			etc.MustStaticFile(etc.SSHConfigResource),
-			trialSSHConfigMode,
-			tar.TypeReg,
-		),
-	}
 	taskToken, err := t.db.StartTaskSession(string(t.task.ID))
 	if err != nil {
 		return errors.Wrap(err, "cannot start a new task session for a trial")
 	}
 	t.preemption = newPreemption()
+
+	trialSpec := &tasks.TrialSpec{
+		Base: *t.taskSpec,
```

Comment:
i might would embed this.. since the fields in Base and the TrialSpec itself should be mutually exclusive (TrialSpec is just a more specific BaseSpec), this would guarantee this at compile time and sort of hint it to the reader.

Reply:
I think embedding a struct looks like inheritance although there is no such concept in GoLang. `TrialSpec` is not just a more specific BaseSpec but makes a `TaskSpec` on the fly. So I would avoid creating any struct relationship between `TaskSpec` and `TrialSpec`.

Reply:
Embedding a struct isn't really inheritance at all, it's composition. A trial spec is trial specific details composed with the base spec, and they are mutual exclusive. I think embedding makes sense here but that's fine.

Reply:
It's inheritance. It inherits all the member functions and you can override any of the functions.

Reply:
It's not inheritance when you just compose two structs. The key detail of inheritance (and what makes it so different/annoying) is the parent-child hierarchy where some methods invoke the parent and child versions, some override, some fall back to the parent and you never know which it will do without digging into the methods themselves. Embedding a struct isn't inheritance but just strictly composition.. the methods on the structs will always do what they always did.. anyway.

Reply:
well, looks like i'm wrong here, they override.. i swear.. give me some time to figure out why i thought this

Reply:
ok at some point i had a refactor with a `tasks.Task` and a `taskspec.Task` and i embedded them and got an error that lead me to believe you couldn't. ok i'm firmly against embedding now, thanks @shiyuann .